compiler/rustc_smir/src/rustc_smir/mod.rs

```diff
@@ -97,34 +97,25 @@ impl<'tcx> Tables<'tcx> {
     where
         F: Fn(&mut Tables<'tcx>, TyCtxt<'tcx>, DefId) -> Option<T>,
     {
-        let crate_def = rustc_span::def_id::DefId { index: CRATE_DEF_INDEX, krate };
+        let crate_def = krate.as_def_id();
         let mut queue = vec![crate_def];
+        queue.extend_from_slice(tcx.trait_impls_in_crate(krate));
         let mut visited = HashSet::default();
         let mut result = vec![];
-        while let Some(next) = queue.pop() {
-            for def_id in tcx.module_children(next).iter().filter_map(|item| item.res.opt_def_id())
-            {
-                if !visited.contains(&def_id) {
-                    visited.insert(def_id);
-                    result.extend(func(self, tcx, def_id));
-                    match tcx.def_kind(def_id) {
-                        DefKind::Mod | DefKind::ForeignMod => queue.push(def_id),
-                        DefKind::Struct | DefKind::Enum | DefKind::Union => {
-                            for associated_item in tcx
-                                .inherent_impls(def_id)
-                                .iter()
-                                .flat_map(|impl_id| tcx.associated_item_def_ids(impl_id))
-                            {
-                                result.extend(func(self, tcx, *associated_item));
-                            }
-                        }
-                        DefKind::Trait => {
-                            for associated_item in tcx.associated_item_def_ids(def_id) {
-                                result.extend(func(self, tcx, *associated_item));
-                            }
-                        }
-                        _ => {}
+        while let Some(def_id) = queue.pop() {
+            if visited.insert(def_id) {
+                result.extend(func(self, tcx, def_id));
+                match tcx.def_kind(def_id) {
+                    DefKind::Mod | DefKind::ForeignMod | DefKind::Trait => queue.extend(
+                        tcx.module_children(def_id).iter().filter_map(|item| item.res.opt_def_id()),
+                    ),
+                    DefKind::Impl { .. } => queue.extend(
+                        tcx.associated_items(def_id).in_definition_order().map(|item| item.def_id),
```

Comment:
I think I mentioned previously using `module_children` for impls and for traits, but I guess I take it back -- can you also move the `DefKInd::Trait` arm here? Using `tcx.associated_items` will also make sure we catch the *synthetic* def ids that correspond to RPITITs/AFITs.

Also, no need to use `associated_items().map()` -- there's an `associated_item_def_ids` query.